docs/layout.rst

```diff
@@ -0,0 +1,246 @@
+    N = 200000
+    pos = ti.var(ti.f32)
+    vel = ti.var(ti.f32)
+    ti.root.dense(ti.i, N).place(pos)
+    ti.root.dense(ti.i, N).place(vel)
+
+    @ti.kernel
+    def step():
+        pos[i] += vel[i] * dt
+        vel[i] += -k * pos[i] * dt
+
+
+Here, we placed ``pos`` and ``vel`` seperately. So the distance in address space between ``pos[i]`` and ``vel[i]`` is ``200000``. This will result in a poor spatial locality and lots of cache-misses, which damages the performance.
+A better placement is to place them together:
+
+.. code-block:: python
+
+    ti.root.dense(ti.i, N).place(pos, vel)
+
+Then ``vel[i]`` is placed right next to ``pos[i]``, this can increases the cache-hit rate and therefore increases performance.
+
+
+Flat layouts versus blocked layouts 
+-------------------------
+
+By default, when allocating a ``ti.var`` , it follows the most naive data layout
+
+.. code-block:: python
+
+  val = ti.var(ti.f32, shape=(32, 64, 128))
+  # C++ equivalent: float val[32][64][128]
+
+However, oftentimes this data layout is suboptimal for computer graphics tasks.
+For example, ``val[i, j, k]`` and ``val[i + 1, j, k]`` are very far away (``32 KB``) from each other, and leads to poor access locality under certain computation tasks. Specifically, in tasks such as texture trilinear interpolation, the two elements are not even within the same ``4KB`` pages, creating a huge cache/TLB pressure.
+
+A better layout might be
+
+.. code-block:: python
+
+  val = ti.var(ti.f32)
+  ti.root.dense(ti.ijk, (8, 16, 32)).dense(ti.ijk, (4, 4, 4)).place(val)
+
+This organizes ``val`` in ``4x4x4`` blocks, so that with high probability ``val[i, j, k]`` and its neighbours are close to each other (i.e., in the same cacheline or memory page).
+
+
+Struct-fors on advanced dense data layouts
+------------------------------------------
+
+Struct-fors on nested dense data structures will automatically follow their data order in memory. For example, if 2D scalar tensor ``A`` is stored in row-major order,
+
+.. code-block:: python
+
+  for i, j in A:
+    A[i, j] += 1
+
+will iterate over elements of ``A`` following row-major order. If ``A`` is column-major, then the iteration follows the column-major order.
+
+If ``A`` is blocked, the iteration will happen within each block first. This maximizes memory bandwidth utilization in most cases.
+
+Struct-fors on sparse tensors follows the same philosophy, and will be discussed further in :ref:`sparse`.
+
+
+Examples
+--------
+
+2D matrix, row-major
+
+.. code-block:: python
+
+  A = ti.var(ti.f32)
+  ti.root.dense(ti.ij, (256, 256)).place(A)
+
+2D matrix, column-major
+
+.. code-block:: python
+
+  A = ti.var(ti.f32)
+  ti.root.dense(ti.ji, (256, 256)).place(A) # Note ti.ji instead of ti.ij
+
+`8x8` blocked 2D array of size `1024x1024`
+
+.. code-block:: python
+
+  density = ti.var(ti.f32)
+  ti.root.dense(ti.ij, (128, 128)).dense(ti.ij, (8, 8)).place(density)
+
+
+3D Particle positions and velocities, arrays-of-structures
```

Comment:
I think you can just write AoS and SoA now.

Reply:
I see, fyi it's written by @yuanming-hu, in the old data_layout.rst.
```suggestion
3D Particle positions and velocities, AoS
```